packages/playwright-test/src/runner.ts

```diff
@@ -378,146 +383,132 @@ export class Runner {
-      if (raceResult.timedOut) {
-        this._reporter.onError?.(createStacklessError(
-            `Timed out waiting ${config.globalTimeout / 1000}s for the global setup to run`));
-        result = { status: 'timedout' } as FullResult;
-      } else {
-        result = raceResult.result;
-      }
-      if (result.status !== 'passed')
-        return result;
+    // Setup the plugins.
+    for (const plugin of this._plugins) {
+      taskRunner.addTask('plugin setup', async () => {
+        await plugin.setup?.(config, config._configDir, this._reporter);
+        return () => plugin.teardown?.();
+      });
     }
 
-    // Run the tests.
-    let fullResult: FullResult;
-    {
-      const remainingTime = deadline - monotonicTime();
-      const raceResult = await raceAgainstTimeout(async () => {
-        try {
-          return await this._innerRun(options);
-        } catch (e) {
-          this._reporter.onError?.(serializeError(e));
-          return { status: 'failed' } as FullResult;
-        } finally {
-          await globalTearDown?.();
-        }
-      }, remainingTime);
-
-      // If timed out, bail.
-      let result: FullResult;
-      if (raceResult.timedOut) {
-        this._reporter.onError?.(createStacklessError(
-            `Timed out waiting ${config.globalTimeout / 1000}s for the entire test run`));
-        result = { status: 'timedout' };
-      } else {
-        result = raceResult.result;
-      }
-
-      // Report end.
-      await this._reporter.onEnd?.(result);
-      fullResult = result;
+    // Run global setup.
+    if (config.globalSetup) {
+      taskRunner.addTask('global setup', async () => {
+        const setupHook = await this._configLoader.loadGlobalHook(config.globalSetup!);
+        const globalSetupResult = await setupHook(this._configLoader.fullConfig());
+        return globalSetupResult && typeof globalSetupResult === 'function' ? globalSetupResult : undefined;
+      });
     }
 
-    // Calling process.exit() might truncate large stdout/stderr output.
-    // See https://github.com/nodejs/node/issues/6456.
-    // See https://github.com/nodejs/node/issues/12921
-    await new Promise<void>(resolve => process.stdout.write('', () => resolve()));
-    await new Promise<void>(resolve => process.stderr.write('', () => resolve()));
-
-    await this._reporter.onExit?.();
-    return fullResult;
-  }
-
-  private async _innerRun(options: RunOptions): Promise<FullResult> {
-    const config = this._configLoader.fullConfig();
-    // Each entry is an array of test groups that can be run concurrently. All
-    // test groups from the previos entries must finish before entry starts.
-    const { rootSuite, testGroups } = await this._collectTestGroups(options);
+    // Schedule teardown.
+    if (config.globalTeardown) {
+      taskRunner.addTask('global teardown', async () => {
+        const teardownHook = await this._configLoader.loadGlobalHook(config.globalTeardown!);
```

Comment:
Let's merge this into 'global setup teardown' phase to avoid running it when global setup was not run.

Reply:
done